### PR TITLE
Changes for indirect launch of SYCL parallel reduce

### PR DESCRIPTION
**Pull request by nliber**
These are updates for SYCL parallel reduce.

### CODE REVIEW

**Comment by masterleinad** [comment]
Which of these commits contain the changes you really want? I am happy to help cleaning this up.

**Comment by masterleinad** [comment]
You will need to fix the indentation.

**Comment by masterleinad** [comment]
After #3480 has been merged, this should be rebased.

**Comment by masterleinad** [comment]
Hmmm... There are a lot of unrelated changes and conflicting files here. Which of the commits do you actually care for? (Maybe just rebase onto upstream/develop).

File: core/unit_test/incremental/Test11a_ParallelFor_TeamThreadRange.hpp
Comment (masterleinad) [inline core/unit_test/incremental/Test11a_ParallelFor_TeamThreadRange.hpp:57]: Is this change related to this pull request? I would move it to a different pull request.

File: core/unit_test/incremental/Test11a_ParallelFor_TeamThreadRange.hpp
Comment (masterleinad) [inline core/unit_test/incremental/Test11a_ParallelFor_TeamThreadRange.hpp:85]: and here

File: core/unit_test/incremental/Test11b_ParallelFor_TeamVectorRange.hpp
Comment (masterleinad) [inline core/unit_test/incremental/Test11b_ParallelFor_TeamVectorRange.hpp:57]: and here

File: core/unit_test/incremental/Test11b_ParallelFor_TeamVectorRange.hpp
Comment (masterleinad) [inline core/unit_test/incremental/Test11b_ParallelFor_TeamVectorRange.hpp:85]: and here

File: core/unit_test/incremental/Test11c_ParallelFor_ThreadVectorRange.hpp
Comment (masterleinad) [inline core/unit_test/incremental/Test11c_ParallelFor_ThreadVectorRange.hpp:58]: adn here

**Review by masterleinad — COMMENTED** [review]
I am mostly fine with this pull request but I am a little concerned about the memory being used after the underlying allocation is freed and whether we can/want to provide some kind of safeguard mechanism.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:232]: FYI, in #3619 it turned out that we really need to run `FunctorFinal` on the device and this requires the functor to possibly be launched indirectly. Hence, I am incorporating the zero-length case into `indirect_launch` again in that pull request.

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:232]: We split the pull request. It's now #3625.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:85]: It'll be slightly more efficient, as `sycl::queue` is a `shared_ptr` under the covers, although it is there because I just tend to add a hidden friend swap function when writing value types more than having given it much thought.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:146]: It is a raw pointer because it has no lifetime to manage.  It could be a unique_ptr with the same deleter though, since calling the pseudo-dtor has no effect.

Will make private.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:163]: Will do.

File: core/unit_test/incremental/Test11a_ParallelFor_TeamThreadRange.hpp
Comment (nliber) [reply core/unit_test/incremental/Test11a_ParallelFor_TeamThreadRange.hpp:57]: Different (I thought it came from a rebase).

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:102]: No.  When the right hand side goes out of scope, it would `sycl::free` the allocation (which the left hand side now refers to) in its destructor.  The left hand side would also have its previous allocation leaked.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:178]: Added an `m_size` member to `USMObjectMem` to track this.

I could make that a std::type_info* instead, which would track the exact type (instead of just its size), but the comparison cost of `&typeid(T) == m_ti || typeid(T) == *m_ti` is potentially expensive (on the off-chance the pointers don't match).

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:209]: No.  capacity can be larger.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:208]: Yeah.  I had it as `const` and then ran into that problem.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:169]: I was mainly concerned that it is annoying if the return type of `copy_from` depends on the function that is called internally and we have to find slightly annoying workarounds to accommodate for both a raw pointer and a `std::unique_ptr`. So I would return a `std::uniqu_ptr<T>` here. I am not feeling strongly about it, though.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:203]: I think this is good enough since it's not user-facing.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:264]: This looks rather weird since `fancy_result_ptr` can either be a raw pointer or a `std::unique_ptr`.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:158]: and here

**Review by masterleinad — COMMENTED** [review]
Overall this looks OK to me after rebasing. I leave it up to you to change the return type of `memcpy_from` to `std::unique_ptr<T>` or not.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:169]: That annoyed me too.

Having given this more thought, I think it is less weird to have a `unique_ptr` which doesn't call the destructor for objects in USM device memory because they are trivially copyable, as well as always tracking whether or not the object in USMObjectMem is live.  I'll make that change.

**Comment by masterleinad** [comment]
This looks good to me but you still need to rebase and fix the formatting.

File: .gitignore
Comment (masterleinad) [inline .gitignore:16]: I'm not quite sure whether we want to add more stuff here.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:152]: Possibly use a Kokkos function like `Kokkos::Impl::throw_runtime_exception()`?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:172]: Possibly catch synchronous errors here and rethrow using `Kokkos::Impl::throw_runtime_exception()`?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:191]: This change looks unrelated.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:244]: also unrelated

**Review by masterleinad — COMMENTED** [review]
Still looks OK to me. Just some comments. We also need someone else looking at this.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:252]: Wasn't  sure how to read your invariants before I read the comment below.  Maybe precede with "all four expressions below must evaluate to `true`"?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:216]: Why do you hav both a return value and an output parameter?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:89]: "detor"?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:142]: Readability
```suggestion
      assert(m_size == 0);
```

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:130]: Might want to `assert(m_q)`

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:114]: What does this do? Reassociating the object with a different queue? After deleting all its stuff?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:127]: so you have to call reset before deleting the object?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (crtrott) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:152]: yeah

Comment (crtrott) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:152]: More importantly, I think we should go through our memory spaces explicitly so that this memory gets tracked by our memory tracking. Or is there a good reason not to do that? I would like internal memory to show up as much as possible in the tools, so folks don't have to ask us where all the extra memory is coming from their apps need. In particular it would allow us to be reasonable confident about what is caused by us vs the runtime of the underlying programming model.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:99]: Is there any good reason to spell out the type?
Rather than just
```C++
auto& indirectKernelMem = instance.m_indirectKernelMem;

**Review by crtrott — CHANGES_REQUESTED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:191]: Undo this

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:244]: Undo

**Review by crtrott — CHANGES_REQUESTED** [review]
Looks largely good. Some questions, and one request - assuming it doesn't have any drawbacks I have not thought about: allocate the memory via memory spaces so the tools can track it.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp]: `copy_from` now always returns a `unique_ptr`.

File: .gitignore
Comment (nliber) [reply .gitignore:16]: gdb.txt is the default filename of the log file generated by gdb option 'set logging on'
.gdb_history is the default filename of the history file generated by the gdb option 'set history save on'.

Other than a minor annoyance, it doesn't matter much to me whether or not these are added to .gdbignore.  Removed.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:191]: Undone

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:244]: Undone.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:127]: It is not necessary to call `reset()` before destructing a `USMObjectMem`.  However, `SYCLInternal::finalize()` resets all its member variables, which is what is supported here.

Note:  it _is_ necessary that there is not an outstanding `unique_ptr` prior to `reset()`/destroying the instance, both because only the `unique_ptr` knows how to destroy what a `USMObjectMem` instance holds and we don't want the memory going away until there are no outstanding pointers or references to it.  Because this is not user-facing, opted for this simpler scheme (with `assert`s) instead of something more complicated and slower using `shared_ptr/weak_ptr`.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:128]: Fixed

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:104]: Fixed

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:252]: Improved.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:89]: Fixed

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:130]: These asserts are really meant as precondition checks on external calls.  In this particular case, what is being checked by the assert is that there not an outstanding `unique_ptr` (`m_data` doesn't point to a live object).  There is no (external) way for `m_q` to not be engaged when `m_data` is set (a bug completely internal to `USMObjectMem` could, of course, cause this, as that would mean the invariants are broken), which is why I don't check it.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:99]: I'm just not a fan of `auto` in general, as it hides too much information from code readers.

Of course, the verbosity of long types also is annoying for code readers...

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:114]: When the `SYCLInternal` object is first constructed, the `USMObjectMem` objects it contains (currently `m_indirectKernelMem` and `m_reductionResultMem`) are default constructed and do not have a sycl::queue associated with them, as we don't as yet have one to give to them.

When we get the queue, we need some way to set it in `m_indirectKernelMem`, etc.  Previous versions just assigned over the old `USMObjectMem` object (`m_indirectKernelMem = IndirectKernelMem(q);`), but assignment of one `sycl::queue` to another (which it did internally) is not working under SYCL+Cuda.  This leaves either changing `m_indirectKernelMem` from being of type `IndirectKernelMem` to type `optional<IndirectKernelMem>` (which requires all uses to dereference it first) or to add a `reset(q)` method.  It is modeled on the `shared_ptr`/`unique_ptr` `.reset()` method.

`reset(...)` is just maintaining the relationship that `m_data` points to memory allocated with `m_q`, so if the old queue goes away, we have to free whatever memory was already allocated first.  While I expect that `reset(q)` is called exactly once (and `m_data` can only be set after `m_q` is engaged, so the first `reset(q)` call doesn't call `sycl::free`), I didn't see any reason to preclude the possibility that it gets called more than once in the future.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:152]: I agree with that, although I'd rather do the change to use memory spaces under a separate P/R, as those changes have to happen to parallel scan as well.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:216]: In debugging it was useful (I could assert the memcpy_to did the right thing in one line).  Would you rather I just make it `void`?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:172]: Agreed.

File: .gitignore
Comment (dalg24) [reply .gitignore:16]: You can use `.git/info/exclude` locally

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:154]: Then mark it with FIXME_SYCL so we don't forget about it.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:180]: Hmm.. I'm not quite sure that this is the right place for it. I would rather go for the execution space `Kokkos::Experimental::SYCL`.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:156]: In hindsight and wrt to #3671, I am not quite sure if we want to treat the result ptr the same as the kernels. In particular, I am not sure if we really want to store the result in `shared` rather than `device` space but I am fine with addressing that later.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp]: I'm trying to do that (I agree that is where it belongs), but ran into incomplete types because of include order.

If its okay, I'll just add a FIXME_SYCL here and work on it separately.

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp]: Yes, this pull request has been dragging long enough. I am in favor of merging and cleaning up when we have merged more of the `SYCL` pull requests.

You still need to rebase apparently, though.

**Comment by masterleinad** [comment]
Hmmm... Looks like something went wrong here. Is this rebased on top of `upstream/develop`?

**Comment by nliber** [comment]
> Hmmm... Looks like something went wrong here. Is this rebased on top of `upstream/develop`?

Yeah, I don't think that worked.  Rebasing again now...

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:190]:
```suggestion
      // id0,id1 encoded within first index; id2,id3 to second index; id4,id5 to
```

**Review by masterleinad — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:239]: Added comment

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:190]: Fixed

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:92]: Fixed

**Comment by masterleinad** [comment]
There is still a conflict here.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:103]: Not sure what I was thinking :)

**Pull request merged**
